File: backend/src/tools/read-file-external.ts
import { readFile, stat } from "node:fs/promises";
import { isAbsolute, resolve } from "node:path";
import { z } from "zod/v4";
import { simpleResult, type Tool, type ToolResult } from "./base";

const MAX_FILE_SIZE = 100_000;

const parameters = z
  .object({
    path: z
      .string()
      .describe(
        "Absolute file path (e.g. /etc/hosts, /home/user/other-project/file.py).",
      ),
    start_line: z
      .number()
      .int()
      .optional()
      .describe(
        "First line to read (1-based, inclusive). Omit to start from the beginning.",
      ),
    end_line: z
      .number()
      .int()
      .optional()
      .describe(
        "Last line to read (1-based, inclusive). Omit to read to the end.",
      ),
  })
  .strict();

type Params = z.infer<typeof parameters>;

export class ReadFileExternalTool implements Tool<typeof parameters> {
  readonly name = "read_file_external";
  readonly description =
    "Read a file anywhere on the filesystem by absolute path. " +
    "Use this when you need to read files outside the project working directory " +
    "(e.g. system config files, files in other projects). " +
    "Requires user approval before execution. " +
    "Returns the file contents with line numbers.";
  readonly parameters = parameters;
  readonly requiresConfirmation = true;

  async execute(args: Params, _workDir: string): Promise<ToolResult> {
    const rawPath = args.path;
    if (!rawPath) {
      return simpleResult("Error: 'path' argument is required.");
    }

    const resolved = resolve(rawPath);

    if (!isAbsolute(rawPath)) {
      return simpleResult(`Error: path '${rawPath}' must be absolute.`);
    }

    let st: Awaited<ReturnType<typeof stat>>;
    try {
      st = await stat(resolved);
    } catch {
      return simpleResult(`Error: file '${rawPath}' does not exist.`);
    }

    if (!st.isFile()) {
      return simpleResult(`Error: '${rawPath}' is not a file.`);
    }

    if (st.size > MAX_FILE_SIZE) {
      return simpleResult(
        `Error: file '${rawPath}' is ${st.size.toLocaleString()} bytes ` +
          `(limit is ${MAX_FILE_SIZE.toLocaleString()} bytes). ` +
          "Use start_line/end_line to read a portion.",
      );
    }

    let text: string;
    try {
      text = await readFile(resolved, "utf-8");
    } catch (exc) {
      return simpleResult(`Error reading '${rawPath}': ${exc}`);
    }

    const lines = text.split("\n");
    if (lines.length > 0 && lines[lines.length - 1] === "") {
      lines.pop();
    }
    const total = lines.length;

    let start = args.start_line ?? 1;
    let end = args.end_line ?? total;

    start = Math.max(1, start);
    end = Math.min(total, end);

    if (start > end) {
      return simpleResult(
        `Error: start_line (${start}) > end_line (${end}). File has ${total} lines.`,
      );
    }

    const selected = lines.slice(start - 1, end);
    const width = String(end).length;
    const numbered = selected.map(
      (line, i) => `${String(start + i).padStart(width)} | ${line}`,
    );
    const header = `File: ${rawPath} (lines ${start}-${end} of ${total})\n`;
    return simpleResult(header + numbered.join("\n"));
  }
}

File: backend-py/src/voxpilot/services/tools/read_file_external.py
"""read_file_external tool — read any file on the filesystem (requires confirmation)."""

from pathlib import Path
from typing import Any, ClassVar

from voxpilot.services.tools.base import Tool

MAX_FILE_SIZE = 100_000  # 100 KB


class ReadFileExternalTool(Tool):
    """Read a file by absolute path, outside the working directory.

    Requires user confirmation before execution.
    """

    name = "read_file_external"
    description = (
        "Read a file anywhere on the filesystem by absolute path. "
        "Use this when you need to read files outside the project working directory "
        "(e.g. system config files, files in other projects). "
        "Requires user approval before execution. "
        "Returns the file contents with line numbers."
    )
    parameters: ClassVar[dict[str, Any]] = {
        "type": "object",
        "properties": {
            "path": {
                "type": "string",
                "description": (
                    "Absolute file path "
                    "(e.g. /etc/hosts, /home/user/other-project/file.py)."
                ),
            },
            "start_line": {
                "type": "integer",
                "description": (
                    "First line to read (1-based, inclusive). "
                    "Omit to start from the beginning."
                ),
            },
            "end_line": {
                "type": "integer",
                "description": "Last line to read (1-based, inclusive). Omit to read to the end.",
            },
        },
        "required": ["path"],
        "additionalProperties": False,
    }

    requires_confirmation = True

    async def execute(self, arguments: dict[str, Any], work_dir: Path) -> str:
        """Read a file by absolute path and return its contents with line numbers."""
        raw_path: str = arguments.get("path", "")
        if not raw_path:
            return "Error: 'path' argument is required."

        resolved = Path(raw_path).resolve()

        if not resolved.is_absolute():
            return f"Error: path '{raw_path}' must be absolute."
        if not resolved.exists():
            return f"Error: file '{raw_path}' does not exist."
        if not resolved.is_file():
            return f"Error: '{raw_path}' is not a file."

        size = resolved.stat().st_size
        if size > MAX_FILE_SIZE:
            return (
                f"Error: file '{raw_path}' is {size:,} bytes "
                f"(limit is {MAX_FILE_SIZE:,} bytes). "
                "Use start_line/end_line to read a portion."
            )

        try:
            text = resolved.read_text(encoding="utf-8", errors="replace")
        except OSError as exc:
            return f"Error reading '{raw_path}': {exc}"

        lines = text.splitlines()
        total = len(lines)

        start: int = arguments.get("start_line", 1)
        end: int = arguments.get("end_line", total)

        start = max(1, start)
        end = min(total, end)

        if start > end:
            return f"Error: start_line ({start}) > end_line ({end}). File has {total} lines."

        selected = lines[start - 1 : end]
        width = len(str(end))
        numbered = [f"{i:{width}d} | {line}" for i, line in enumerate(selected, start=start)]
        header = f"File: {raw_path} (lines {start}-{end} of {total})\n"
        return header + "\n".join(numbered)
